lib/Platform/Unicode/CharacterProperties.cpp

```diff
@@ -210,4 +211,148 @@ uint32_t canonicalize(uint32_t cp, bool unicode) {
+  auto rangeMapStart = std::begin(unicodePropertyRangeMap_BinaryProperty);
+  auto rangeMapEnd = std::end(unicodePropertyRangeMap_BinaryProperty);
+  auto nameMapStart = std::begin(canonicalPropertyNameMap_BinaryProperty);
+  auto nameMapEnd = std::end(canonicalPropertyNameMap_BinaryProperty);
+
+  if (propertyValue.empty()) {
+    // There was no property value, this is either a binary property or a value
+    // from General_Category, as per `LoneUnicodePropertyNameOrValue`.
+    if (findNameMapEntry(nameMapStart, nameMapEnd, key) == nameMapEnd) {
+      rangeMapStart = std::begin(unicodePropertyRangeMap_GeneralCategory);
+      rangeMapEnd = std::end(unicodePropertyRangeMap_GeneralCategory);
+      nameMapStart = std::begin(canonicalPropertyNameMap_GeneralCategory);
+      nameMapEnd = std::end(canonicalPropertyNameMap_GeneralCategory);
+    }
+  } else {
+    // There was a property value, assume the name is a category.
+    key = propertyValue;
+    if (propertyName == "General_Category" || propertyName == "gc") {
+      rangeMapStart = std::begin(unicodePropertyRangeMap_GeneralCategory);
+      rangeMapEnd = std::end(unicodePropertyRangeMap_GeneralCategory);
+      nameMapStart = std::begin(canonicalPropertyNameMap_GeneralCategory);
+      nameMapEnd = std::end(canonicalPropertyNameMap_GeneralCategory);
+    } else if (propertyName == "Script" || propertyName == "sc") {
+      rangeMapStart = std::begin(unicodePropertyRangeMap_Script);
+      rangeMapEnd = std::end(unicodePropertyRangeMap_Script);
+      nameMapStart = std::begin(canonicalPropertyNameMap_Script);
+      nameMapEnd = std::end(canonicalPropertyNameMap_Script);
+    } else if (propertyName == "Script_Extensions" || propertyName == "scx") {
+      // Script_Extensions is a superset of Script.
+      if (!addUnicodePropertyRanges(receiver, "Script", propertyValue)) {
+        return false;
+      }
+      rangeMapStart = std::begin(unicodePropertyRangeMap_ScriptExtensions);
+      rangeMapEnd = std::end(unicodePropertyRangeMap_ScriptExtensions);
+      nameMapStart = std::begin(canonicalPropertyNameMap_Script);
+      nameMapEnd = std::end(canonicalPropertyNameMap_Script);
+    } else {
+      return false;
+    }
+  }
+
+  // Canonicalize the property name.
+  auto canonicalNameEntry = findNameMapEntry(nameMapStart, nameMapEnd, key);
+  if (canonicalNameEntry == nameMapEnd) {
+    return false;
+  }
+
+  // Look up the range arrays for the property.
+  auto rangeMapEntry = findRangeMapEntry(
+      rangeMapStart, rangeMapEnd, canonicalNameEntry->canonical);
+  if (rangeMapEntry == rangeMapEnd) {
+    return false;
+  }
+
+  auto rangeArrayPoolStart = std::next(
+      std::begin(UNICODE_RANGE_ARRAY_POOL),
+      rangeMapEntry->rangeArrayPoolOffset);
+  auto rangeArrayPoolEnd =
+      std::next(rangeArrayPoolStart, rangeMapEntry->rangeArraySize);
+
+  for (auto rangePoolRef = rangeArrayPoolStart;
+       rangePoolRef != rangeArrayPoolEnd;
+       ++rangePoolRef) {
```

Comment:
Similarly here, it may make things simpler to use `ArrayRef`
```suggestion
  llvh::ArrayRef rangeArrayPool{&UNICODE_RANGE_ARRAY_POOL[rangeMapEntry->rangeArrayPoolOffset], rangeMapEntry->rangeArraySize};

  for (auto rangePoolRef: rangeArrayPool) {
```